fearless/Modules/Staking/StakingMain/StakingMainInteractor.swift

```diff
@@ -19,10 +40,91 @@ final class StakingMainInteractor {
+                } catch {
+                    self?.presenter.didReceive(calculatorError: error)
+                }
+            }
+        }
+
+        operationManager.enqueue(operations: [operation],
+                                 in: .transient)
+    }
+
+    private func subscribeToPriceChanges() {
+        let updateClosure = { [weak self] (changes: [DataProviderChange<PriceData>]) in
+            if changes.isEmpty {
+                self?.presenter.didReceive(price: nil)
+            } else {
+                for change in changes {
+                    switch change {
+                    case .insert(let item), .update(let item):
+                        self?.presenter.didReceive(price: item)
+                    case .delete:
+                        self?.presenter.didReceive(price: nil)
+                    }
+                }
+            }
+        }
+
+        let failureClosure = { [weak self] (error: Error) in
+            self?.presenter.didReceive(error: error)
+            return
+        }
+
+        let options = DataProviderObserverOptions(alwaysNotifyOnRefresh: false,
+                                                  waitsInProgressSyncOnAdd: false)
+        priceProvider.addObserver(self,
+                                  deliverOn: .main,
+                                  executing: updateClosure,
+                                  failing: failureClosure,
+                                  options: options)
+    }
+
+    private func subscribeToAccountChanges() {
+        let updateClosure = { [weak self] (changes: [DataProviderChange<DecodedAccountInfo>]) in
+            if changes.isEmpty {
+                self?.presenter.didReceive(balance: nil)
+            } else {
+                for change in changes {
+                    switch change {
+                    case .insert(let wrapped), .update(let wrapped):
+                        self?.presenter.didReceive(balance: wrapped.item.data)
+                    case .delete:
+                        self?.presenter.didReceive(price: nil)
+                    }
+                }
+            }
+        }
+
+        let failureClosure = { [weak self] (error: Error) in
+            self?.presenter.didReceive(error: error)
+            return
+        }
+
+        let options = DataProviderObserverOptions(alwaysNotifyOnRefresh: false,
+                                                  waitsInProgressSyncOnAdd: false)
+        balanceProvider.addObserver(self,
+                                    deliverOn: .main,
+                                    executing: updateClosure,
+                                    failing: failureClosure,
+                                    options: options)
+    }
 }
 
 extension StakingMainInteractor: StakingMainInteractorInputProtocol {
     func setup() {
+        subscribeToPriceChanges()
+        subscribeToAccountChanges()
+        provideRewardCalculator()
         eventCenter.add(observer: self, dispatchIn: .main)
```

Comment:
Looks like we need to add handling for account, connection and runtime changes and provide new calculator to presenter.